iam-token-manager/v1.ts

```diff
@@ -0,0 +1,198 @@
+/**
+ * Copyright 2015 IBM Corp. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import extend = require('extend');
+import { sendRequest } from '../lib/requestwrapper';
+
+export type Options = {
+  iam_apikey?: string;
+  iam_access_token?: string;
+  iam_url?: string;
+}
+
+export interface IamTokenData {
+  access_token: string;
+  refresh_token: string;
+  token_type: string;
+  expires_in: number;
+  expiration: number;
+}
+
+export class IamTokenManagerV1 {
+  name: string;
+  serviceVersion: string;
+  protected url: string;
+  protected tokenInfo: IamTokenData;
+  private iam_apikey: string; // tslint:disable-line variable-name
+  private userAccessToken: string;
+
+  /**
+   * IAM Token Manager Service
+   *
+   * Retreives, stores, and refreshes IAM tokens.
+   *
+   * @param {Object} options
+   * @param {String} options.iam_apikey
+   * @param {String} options.iam_access_token
+   * @param {String} options.iam_url - url of the iam api to retrieve tokens from
+   * @constructor
+   */
+  constructor(options: Options) {
+    this.url = options.iam_url || 'https://iam.ng.bluemix.net/identity/token';
+    this.tokenInfo = {} as IamTokenData;
+    if (options.iam_apikey) {
+      this.iam_apikey = options.iam_apikey;
+    }
+    if (options.iam_access_token) {
+      this.userAccessToken = options.iam_access_token;
+    }
+  }
+
+  /**
+   * This function sends an access token back through a callback. The source of the token
+   * is determined by the following logic:
+   * 1. If user provides their own managed access token, assume it is valid and send it
+   * 2. If this class is managing tokens and does not yet have one, make a request for one
+   * 3. If this class is managing tokens and the token has expired, refresh it
+   * 4. If this class is managing tokens and has a valid token stored, send it
+   *
+   * @param {Function} cb - callback function that the token will be passed to
+   */
+  public getToken(cb: Function) {
+    if (this.userAccessToken) {
+      // 1. use user-managed token
+      return cb(null, this.userAccessToken);
+    } else if (!this.tokenInfo.access_token) {
+      // 2. request an initial token
```

Comment:
two and tree can be merged.

```js
if (!this.tokenInfo.access_token || this.isTokenExpired()) {
  //...
}
```

Reply:
The separation exists here because for `2` the token is _requested_ but for `3` the token is _refreshed_. While they are pretty similar, I think it is cleaner to leave them separate. Let me know what you think

Reply:
my bad.